### PR TITLE
feat: Add trigger heuristic grader

### LINKED ISSUE #80: feat: Add trigger heuristic grader
## Summary

Add a new grader that validates whether a skill should or shouldn't activate for a given prompt. The grader extracts keywords from SKILL.md (name, description, content) and matches them against the task prompt with a configurable threshold.

## Motivation

Currently there's no way to validate trigger accuracy in evals — whether a skill correctly activates for relevant prompts and stays silent for irrelevant ones. This is the primary missing grader type for skill quality assessment.

## Proposed Implementation

Create `internal/graders/trigger_grader.go`:

- **Inputs:** SKILL.md path, prompt text, expected outcome (should_trigger / should_not_trigger)
- **Keyword extraction:** Parse SKILL.md description, trigger phrases, and content to build a keyword set
- **Matching:** Score prompt against keyword set using configurable threshold (default: 0.6)
- **Modes:**
  - `positive`: skill SHOULD activate for the prompt (score >= threshold → pass)
  - `negative`: skill should NOT activate (score < threshold → pass)

## Example eval.yaml

```yaml
graders:
  - type: trigger
    params:
      skill_path: skills/azure-deploy/SKILL.md
      mode: positive
      threshold: 0.6
```

## Acceptance Criteria

- [ ] `trigger_grader.go` with keyword extraction from SKILL.md
- [ ] Configurable threshold and positive/negative modes
- [ ] Tests covering: positive match, negative match, threshold edge cases
- [ ] Registered in grader factory

### PR DESCRIPTION
**Pull request by spboyer**
Closes #80

### CODE REVIEW

File: internal/graders/trigger_grader.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+		if _, isStopWord := stopWords[field]; isStopWord {
+			continue
+		}
+		tokens = append(tokens, field)
+	}
+	return tokens
+}
+
+var stopWords = map[string]struct{}{
+	"the": {}, "and": {}, "for": {}, "with": {}, "this": {}, "that": {}, "from": {},
```
Comment (wbreza) [inline internal/graders/trigger_grader.go:270]: **[Medium]** Stop words include use, when, and help which are also common trigger verbs. Consider removing these from stop words in trigger scoring context, as phrase matching only partially compensates.

File: internal/graders/trigger_grader_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+description: "Deploy to Azure resources. USE FOR: deploy to azure, publish api, release to cloud. DO NOT USE FOR: write unit tests."
+---
+# Azure deploy skill
+
+## Deployment
+Use this skill for Azure deployment workflows.
+`
+	require.NoError(t, os.WriteFile(path, []byte(content), 0o644))
+	return path
+}
```
Comment (wbreza) [inline internal/graders/trigger_grader_test.go:174]: **[Low]** Nil grading context path untested

`Grade()` gracefully handles nil `gradingContext` and nil `TestCase` (trigger_grader.go lines 182-190), but no test exercises this path. Consider adding a quick test to verify the nil-safety behavior.

File: internal/graders/trigger_grader.go (hunk 1 above)
Comment (wbreza) [inline internal/graders/trigger_grader.go:270]: **[Low]** Prompt tokenized twice

`tokenize(prompt)` is called here and again in `scorePrompt()` (line 237). The prompt tokens could be computed once and passed through. Very minor — only matters in batch scoring.

File: internal/suggest/suggest.go [hunk 3]
```diff
@@ -188,6 +188,7 @@ func AvailableGraderTypes() []string {
 		string(models.GraderKindBehavior),
 		string(models.GraderKindActionSequence),
 		string(models.GraderKindSkillInvocation),
+		string(models.GraderKindTrigger),
 		string(models.GraderKindDiff),
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/suggest/suggest.go:192]: Follow-up scope — AvailableGraderTypes() omitting tool_constraint is a separate concern from the trigger grader PR. Filed mentally for future work.

File: docs/graders/trigger.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# `trigger` Grader
+
+Heuristic grader for validating whether a prompt should activate a skill.
+
+## Use Cases
+
```
[the remark this answers is not in the document]
Comment (spboyer) [reply docs/graders/trigger.md]: Doc formatting is Saul's domain. Deferring to doc review.

**Review by richardpark-msft — APPROVED** [review]
You have a conflict to resolve, but copilot should be able to pick up the pattern. 

 LGTM, is our plan to do some more inference here (like, embeddings, etc), or will it always be a kind of "literal" match against keywords?

File: internal/graders/grader.go [hunk 5]
```diff
@@ -76,6 +76,12 @@ func Create(identifier string, params models.GraderParameters) (Grader, error) {
 		return NewJSONSchemaGrader(identifier, p)
 	case models.ProgramGraderParameters:
 		return NewProgramGrader(identifier, p)
+	case models.TriggerHeuristicGraderParameters:
+		return NewTriggerHeuristicGrader(identifier, TriggerHeuristicGraderParams{
```
Comment (richardpark-msft) [inline internal/graders/grader.go]: Can just pass 'p' and get rid of the Params struct.

File: internal/graders/trigger_grader.go [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+type triggerHeuristicGrader struct {
+	name           string
+	mode           triggerHeuristicMode
+	threshold      float64
+	skillPath      string
+	keywords       map[string]struct{}
+	triggerPhrases []string
+}
+
+func NewTriggerHeuristicGrader(name string, params TriggerHeuristicGraderParams) (*triggerHeuristicGrader, error) {
```
Comment (richardpark-msft) [inline internal/graders/trigger_grader.go]:
```suggestion
func NewTriggerHeuristicGrader(name string, params models.TriggerHeuristicGraderParameters) (*triggerHeuristicGrader, error) {
```

File: internal/graders/trigger_grader.go [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+	"github.com/microsoft/waza/internal/models"
+	"github.com/microsoft/waza/internal/scaffold"
+	"github.com/microsoft/waza/internal/skill"
+)
+
+const defaultTriggerThreshold = 0.6
+
+// TriggerHeuristicGraderParams configures the trigger heuristic grader.
+type TriggerHeuristicGraderParams struct {
```
Comment (richardpark-msft) [inline internal/graders/trigger_grader.go]: Delete this in favor of the one in `models`.

**Review by richardpark-msft — APPROVED** [review]
Still looks fine to me. Added some optional comments, which we can pick off in another PR so this one can get merged.

**Review by wbreza — APPROVED** [review]
## Re-Review: PR #90 — feat: Add trigger heuristic grader

All previous review feedback has been properly addressed:

✅ **Stop words updated** — Removed 'when', 'use', 'help' to preserve trigger verbs  
✅ **Double tokenization eliminated** — estPhraseScore now receives pre-computed tokens  
✅ **Nil context handling added** — TestTriggerHeuristicGrader_NilGradingContext covers both nil context and nil TestCase  
✅ **Empty prompt guard added** — Returns clear feedback when prompt is empty  
✅ **Unique token normalization** — Score calculation divides by unique token count to prevent dilution  
✅ **Comprehensive test coverage** — Including duplicate token test verifying normalization works correctly

The implementation is clean, well-tested, and follows existing patterns. All tests pass.

**Approval:** ✅ Ready to merge
